### PR TITLE
Have ability to restore intial state with expectation loaded from JSON files

### PR DESCRIPTION
**Pull request by drudoi**
Issue link: https://github.com/mcustiel/phiremock/issues/27

### CODE REVIEW

File: src/Server/Actions/RestoreExpectationsAction.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * {@inheritdoc}
+     *
+     * @see \Mcustiel\PowerRoute\Actions\ActionInterface::execute()
+     */
+    public function execute(TransactionData $transactionData, $argument = null)
+    {
+        $this->expectationStorage->clearExpectations();
+        $this->requestStorage->clearRequests();
+        $this->scenarioStorage->clearScenarios();
```
Comment (drudoi) [inline src/Server/Actions/RestoreExpectationsAction.php:58]: I think the user wants to have clean behaviour when initial expectations are restored, and it's quicker do in one request that send 2 more. Anyway, it passes our needs as we want to have test execution so quick as possible.

File: src/Server/Utils/FileExpectationsLoader.php [hunk 2]
```diff
@@ -52,6 +58,10 @@ public function loadExpectationFromFile($fileName)
 
         $this->logger->debug('Parsed expectation: ' . var_export($expectation, true));
         $this->storage->addExpectation($expectation);
+        // As we have no API to modify expectation parsed the same object could be used for backup.
+        // On futher changes when $expectation modifications are possible something like deep-copy
+        // should be used to clone expectation.
+        $this->backup->addExpectation($expectation);
```
Comment (drudoi) [inline src/Server/Utils/FileExpectationsLoader.php:64]: It's quicker to backup expectations than re-parse them from files.

File: src/Server/Actions/RestoreExpectationsAction.php (hunk 1 above)
Comment (mcustiel) [reply src/Server/Actions/RestoreExpectationsAction.php:58]: As I user I would expect this method to reload the expectation from the files. But it is cleaning the scenarios, requests and expectations. Those are unexpected secondary effects. It should only reload the files.

File: src/Server/Utils/FileExpectationsLoader.php (hunk 2 above)
Comment (mcustiel) [reply src/Server/Utils/FileExpectationsLoader.php:64]: The amount of files will never be too high as to be a performance issue when reloading, and we avoid adding complexity and data duplication in memory.

**Review by mcustiel — CHANGES_REQUESTED** [review]

File: src/Server/Utils/FileExpectationsLoader.php (hunk 2 above)
Comment (drudoi) [reply src/Server/Utils/FileExpectationsLoader.php:64]: We would like to have all expectations defined in files to share it beetween tests for different modules in own repos, we have high count of files.

File: src/Server/Actions/RestoreExpectationsAction.php (hunk 1 above)
Comment (drudoi) [reply src/Server/Actions/RestoreExpectationsAction.php:58]: I'll change issue and add description there to describe our needs

Comment (mcustiel) [reply src/Server/Actions/RestoreExpectationsAction.php:58]: Hey. It was not about the description, but the side effects of the feature. If I am restoring the expectations from json I want to only do that. In the future some other user of phiremock could have the case of having a expectation created through the api and then he wants to reload the json files and the other expectation is deleted or the list of requests is deleted and he did not want that, he just wanted the json files to be reloaded. So it is a side effect.
For me the best option is just reload and nothing else, the problem is to identify the expectations loaded from JSON files to only overwrite them, I will have a look at it as soon as I have some free time.

Comment (drudoi) [reply src/Server/Actions/RestoreExpectationsAction.php:58]: Look for https://github.com/mcustiel/phiremock-codeception-extension/pull/20 please, it allows to provide cleanup in one call and saves ~0.2-0.4s for each test in comparation with 3 calls send for cleanup (clean counters, clean expectations, load JSON defined), with 1000 tests (its real number for a project) it saves 3-6 minutes.

Comment (mcustiel) [reply src/Server/Actions/RestoreExpectationsAction.php:58]: Yes. I understand the advantage in performance. 
My proposal is to have a separate url for this feature. Something like `__phiremock/reset`. By sending a post or get request to it we reset everything to initial state by reseting expectations, scenarios and requests counter, and reloading json files. What do you think? 
Later we can add the put request to expectations that reloads the mappings only.

**Pull request merged**